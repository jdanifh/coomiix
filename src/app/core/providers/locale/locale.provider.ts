import { importProvidersFrom, LOCALE_ID } from '@angular/core';
import { HttpBackend, HttpClient } from '@angular/common/http';
import { MAT_DATE_LOCALE } from '@angular/material/core';
import { MAT_MOMENT_DATE_FORMATS, provideMomentDateAdapter } from '@angular/material-moment-adapter';
import {TranslateModule, TranslateLoader} from "@ngx-translate/core";
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { registerLocaleData } from '@angular/common';

// EN Locale
import en from '@angular/common/locales/en';
registerLocaleData(en);
// ES Locale
import 'moment/locale/es';
import es from '@angular/common/locales/es';
registerLocaleData(es);

export enum LOCALES {
  EN = 'en-US',
  ES = 'es-ES'
}
export const DEFAULT_LOCALE = LOCALES.EN;

export const provideLocales = () => {
  return [
    {provide: LOCALE_ID, useValue: DEFAULT_LOCALE},
    {provide: MAT_DATE_LOCALE, useValue: DEFAULT_LOCALE},
    provideMomentDateAdapter(MAT_MOMENT_DATE_FORMATS, {useUtc: true}),
    importProvidersFrom([TranslateModule.forRoot({
      defaultLanguage: DEFAULT_LOCALE,
      loader: {
        provide: TranslateLoader,
        deps: [HttpBackend],
        useFactory: (http: HttpBackend) => new TranslateHttpLoader(new HttpClient(http), './i18n/', '.json'),
      },
    })])
  ]
};

export const provideLocalesTest = () => {
  return [
    provideLocales(),
    { provide: HttpBackend, useClass: HttpBackend },
  ]
}
