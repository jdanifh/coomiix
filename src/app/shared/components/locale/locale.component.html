<button [matMenuTriggerFor]="menu" mat-icon-button aria-label="change locale">
  <mat-icon aria-hidden="true">language</mat-icon>
</button>
<mat-menu #menu="matMenu">
  @for (locale of locales; track locale) {
    <button (click)="changeLocale(locale)" mat-menu-item>
      <span class="locale">{{ locale | uppercase | slice:0:2 }}</span>
      <span>{{ locale }}</span>
    </button>
  }
</mat-menu>
