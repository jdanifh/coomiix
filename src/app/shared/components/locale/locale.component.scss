:host {
  margin-left: auto;
  margin-right: 25px;
}

.locale {
  font-weight: 700;
  margin-right: 25px;
}
