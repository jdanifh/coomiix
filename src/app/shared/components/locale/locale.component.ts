import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatMenuModule } from '@angular/material/menu';
import { LocaleService } from './locale.service';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'cmx-locale',
  imports: [
    CommonModule,
    MatIconModule,
    MatButtonModule,
    MatMenuModule,
  ],
  templateUrl: './locale.component.html',
  styleUrl: './locale.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class LocaleComponent {

  private _service = inject(LocaleService);

  public locale = this._service.locale;
  public locales = this._service.locales;

  public changeLocale(locale: string) {
    this._service.changeLocale(locale);
  }

}
