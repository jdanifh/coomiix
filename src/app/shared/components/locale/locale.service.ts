import { inject, Injectable, signal } from '@angular/core';
import { DateAdapter } from '@angular/material/core';
import { TranslateService } from '@ngx-translate/core';
import { DEFAULT_LOCALE, LOCALES } from '../../../core/providers/locale/locale.provider';

export const LOCALE_KEY = 'LOCALE_ID';

@Injectable({
  providedIn: 'root'
})
export class LocaleService {

  private _locale = signal<string>(DEFAULT_LOCALE);
  private _adapter = inject(DateAdapter);
  private _service = inject(TranslateService);

  constructor() {
    this._service.addLangs(Object.values(LOCALES));
    const savedLocale = localStorage.getItem(LOCALE_KEY);
    if (savedLocale) {
      this.changeLocale(savedLocale);
    } else {
      this.changeLocale(DEFAULT_LOCALE);
    }
  }

  public get locale() {
    return this._locale.asReadonly();
  }

  public get locales() {
    return this._service.getLangs();
  }

  public changeLocale(locale: string) {
    this._locale.set(locale);
    this._service.use(locale);
    this._adapter.setLocale(locale);
    localStorage.setItem(LOCALE_KEY, locale);
  }
}
