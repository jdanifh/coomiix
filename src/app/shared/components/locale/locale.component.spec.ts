import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LocaleComponent } from './locale.component';
import { DEFAULT_LOCALE, LOCALES, provideLocales } from '../../../core/providers/locale/locale.provider';
import { HttpBackend } from '@angular/common/http';

describe('LocaleComponent', () => {
  let component: LocaleComponent;
  let fixture: ComponentFixture<LocaleComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [LocaleComponent],
      providers: [
        provideLocales(),
        { provide: HttpBackend, useClass: HttpBackend },
      ]
    })
    .compileComponents();

    fixture = TestBed.createComponent(LocaleComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render DEFAULT_LOCALE', () => {
    expect(component.locale()).toBe(DEFAULT_LOCALE);
  });

  it('should render locale list', () => {
    expect(component.locales).toStrictEqual(Object.values(LOCALES));
  });

  it('should change current locale', () => {
    expect(component.locale()).toBe(DEFAULT_LOCALE);
    component.changeLocale(LOCALES.ES);
    expect(component.locale()).toBe(LOCALES.ES);
  });

  it(`should match snapshot`, () => {
    expect(fixture.nativeElement.innerHTML).toMatchSnapshot();
  })

});
