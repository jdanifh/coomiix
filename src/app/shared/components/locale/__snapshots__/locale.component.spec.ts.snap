// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleComponent should match snapshot 1`] = `"<button mat-icon-button="" aria-label="change locale" class="mat-mdc-menu-trigger mdc-icon-button mat-mdc-icon-button mat-unthemed mat-mdc-button-base" mat-ripple-loader-uninitialized="" mat-ripple-loader-class-name="mat-mdc-button-ripple" mat-ripple-loader-centered="" ng-reflect-menu="[object Object]" aria-haspopup="menu" aria-expanded="false"><span class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"></span><mat-icon role="img" aria-hidden="true" class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color" data-mat-icon-type="font">language</mat-icon><span class="mat-focus-indicator"></span><span class="mat-mdc-button-touch-target"></span></button><!--container--><mat-menu><!--container--></mat-menu>"`;
